ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/SSLEngineBufferAllocator.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.tls
+
+import java.nio.*
+import javax.net.ssl.*
+
+internal class SSLEngineBufferAllocator(private val engine: SSLEngine) {
+    private var packetBufferSize = 0
+    private var applicationBufferSize = 0
+
+    fun allocatePacket(length: Int): ByteBuffer = allocate(
+        length,
+        get = { packetBufferSize },
+        set = { packetBufferSize = it },
+        new = { engine.session.packetBufferSize }
+    )
+
+    fun allocateApplication(length: Int): ByteBuffer = allocate(
+        length,
+        get = { applicationBufferSize },
+        set = { applicationBufferSize = it },
+        new = { engine.session.applicationBufferSize }
+    )
+
+    fun reallocatePacket(buffer: ByteBuffer, flip: Boolean): ByteBuffer =
+        reallocate(buffer, flip, ::allocatePacket)
+
+    fun reallocateApplication(buffer: ByteBuffer, flip: Boolean): ByteBuffer =
+        reallocate(buffer, flip, ::allocateApplication)
+
+    private inline fun allocate(
+        length: Int,
+        get: () -> Int,
+        set: (Int) -> Unit,
+        new: () -> Int
+    ): ByteBuffer = synchronized(this) {
+        if (get() == 0) {
+            set(new())
+        }
+        if (length > get()) {
+            set(length)
+        }
+        return ByteBuffer.allocate(get())
+    }
+
+    private inline fun reallocate(
+        buffer: ByteBuffer,
+        flip: Boolean,
+        allocate: (length: Int) -> ByteBuffer
+    ): ByteBuffer {
```

Comment:
Could we just make large enough buffers upfront and use a regular pool? the maximum packet size is limited by the protocol

Reply:
sslengine max buffer size is 16K, and we need up to 4 such buffers for every TLS socket implementation, so 64K per socket.
For client side - it's ok, as there will not so much active sockets at the same time, but for server side it can cause memory issues with a lot of clients.
Note: most of the time, those 16K are not fully used, and buffer of smaller size is enough to perform TLS operations